internal/controller/provider_controller.go

```diff
@@ -0,0 +1,104 @@
+// Copyright Axis Communications AB.
+//
+// For a full list of individual contributors, please see the commit history.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controller
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	etosv1alpha1 "github.com/eiffel-community/etos/api/v1alpha1"
+)
+
+// ProviderReconciler reconciles a Provider object
+type ProviderReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=etos.eiffel-community.github.io,resources=providers,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=etos.eiffel-community.github.io,resources=providers/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=etos.eiffel-community.github.io,resources=providers/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.18.4/pkg/reconcile
+func (r *ProviderReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	provider := &etosv1alpha1.Provider{}
+	err := r.Get(ctx, req.NamespacedName, provider)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.Info("provider not found. ignoring object")
+			return ctrl.Result{}, nil
+		}
+		logger.Error(err, "failed to get provider")
+		return ctrl.Result{}, err
+	}
+	logger.V(2).Info("Checking availability of provider", "provider", req.NamespacedName)
+
+	// TODO: Schedule checks instead of doing it every time something happens
+
+	// We don't check the availability of JSONTas as it is not yet running as a service we can check.
+	if (etosv1alpha1.JSONTas{}) == provider.Spec.JSONTas {
+		resp, err := http.Get(fmt.Sprintf("%s/%s", provider.Spec.Host, provider.Spec.Healthcheck.Endpoint))
+		if err != nil {
```

Comment:
Should retries be done here? Or is it handled by Kubernetes outside this controller?

Reply:
This Reconciliation loop is run, at least, every `provider.Spec.Healthcheck.IntervalSeconds`(default 30) seconds.